chiron/integrator.py

```diff
@@ -0,0 +1,135 @@
+            Box vectors for periodic boundary conditions.
+        progress_bar : bool, optional
+            Flag indicating whether to display a progress bar during integration.
+        """
+        from .utils import get_list_of_mass
+
+        self.box_vectors = box_vectors
+        self.progress_bar = progress_bar
+        self.potential = potential
+        self.velocities = None
+
+        self.kB = unit.BOLTZMANN_CONSTANT_kB * unit.AVOGADRO_CONSTANT_NA
+        self.mass = get_list_of_mass(topology)
+
+    def set_velocities(self, vel: unit.Quantity):
+        self.velocities = vel
+
+    def run(
+        self,
+        x0,
+        temperature: unit.Quantity,
+        n_steps: int = 5_000,
+        stepsize=1.0 * unit.femtoseconds,
+        collision_rate=1.0 / unit.picoseconds,
+        key=random.PRNGKey(0),
+    ) -> Dict[str, jnp.array]:
+        """
+        Run the integrator to perform Langevin dynamics molecular dynamics simulation.
+
+        Parameters
+        ----------
+        x0 : array_like
+            Initial positions of the particles.
+        temperature : unit.Quantity
+            Temperature of the system.
+        n_steps : int, optional
+            Number of simulation steps to perform.
+        stepsize : unit.Quantity, optional
+            Time step size for the integration.
+        collision_rate : unit.Quantity, optional
+            Collision rate for the Langevin dynamics.
+        key : jax.random.PRNGKey
+            Random key for generating random numbers.
+
+        Returns
+        -------
+        list of array_like
+            Trajectory of particle positions at each simulation step.
+        """
+
+        log.info("Running Langevin dynamics")
+        log.info(f"n_steps = {n_steps}")
+        log.info(f"stepsize = {stepsize}")
+        log.info(f"collision_rate = {collision_rate}")
+        log.info(f"temperature = {temperature}")
+
+        kbT_unitless = (self.kB * temperature).value_in_unit_system(unit.md_unit_system)
+        mass_unitless = jnp.array(self.mass.value_in_unit_system(unit.md_unit_system))
+        sigma_v = jnp.sqrt(kbT_unitless / mass_unitless)
+        stepsize_unitless = stepsize.value_in_unit_system(unit.md_unit_system)
+        collision_rate_unitless = collision_rate.value_in_unit_system(
+            unit.md_unit_system
+        )
+
+        # Initialize velocities
+        if self.velocities is None:
+            v0 = sigma_v * random.normal(key, x0.shape)
+        else:
+            v0 = self.velocities.value_in_unit_system(unit.md_unit_system)
+        # Convert to dimensionless quantities
+        a = jnp.exp((-collision_rate_unitless * stepsize_unitless))
+        b = jnp.sqrt(1 - jnp.exp(-2 * collision_rate_unitless * stepsize_unitless))
+
+        x = x0.value_in_unit_system(unit.md_unit_system)
+        v = v0
+
+        traj = [x]
+        energy = [self.potential.compute_energy(x)]
+
+        random_noise_v = random.normal(key, (n_steps, x.shape[-1]))
```

Comment:
I'm not sure if it would be better to generate random_noise in the loop (where we use random.split to get new subkeys); it would avoid generating large arrays if we had larger systems and long trajectories. the subkey could be returned along with position, energy, velocity, etc. for reproducibility.

Reply:
It seemed as if this had slowed down the timing of the calculation. Originally it was done inside the loop, and this is not something we can do for larger systems.